### PR TITLE
Topic/eq

### PR DESCRIPTION
**Pull request by bturrubiates**
Not ready for merging. For review once again.
@sungeunchoi @hppritcha

### CODE REVIEW

**Comment by hppritcha** [comment]
:+1: 

looks good

**Comment by bturrubiates** [comment]
Okay, I eventually need to get an EQ test together...

File: prov/gni/src/gnix_eq.c [hunk 1]
```diff
@@ -170,22 +241,195 @@ static int gnix_eq_close(struct fid *fid)
...
+	 * - Handle fastlock_acquire failure case.
+	 */
+	fastlock_acquire(&gnix_eq->lock);
+
+	/*
+	 * Insert generated entry into the event queue.
+	 */
+	slist_insert_tail(&ev_entry->entry, &gnix_eq->ev_queue);
+
+	fastlock_release(&gnix_eq->lock);
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_eq.c]: Food for thought: Might consider a wrapper for the set of these 3 operations (acquire, insert, release) with the error check if it become more common.

**Comment by sungeunchoi** [comment]
:+1:

**Comment by bturrubiates** [comment]
Yeah, actually I kind of want to make a read/write function that is generic and have both error and event use it. Just need to get started on that so it's a bit cleaner.

**Comment by bturrubiates** [comment]
Fixed mistakes I couldn't catch before because I forgot to add it to compiling sources...

Tried it with the current eq_test.c in fabtests and passes 3/5 tests. Want to do a bit of cleanup, but that can come later.

@sungeunchoi @hppritcha

**Comment by hppritcha** [comment]
so should this be merged now?

**Comment by bturrubiates** [comment]
@hppritcha, no I just found a mistake.

**Comment by bturrubiates** [comment]
@sungeunchoi @hppritcha I did some cleanup. Look it over please.

File: prov/gni/src/gnix_eq.c [hunk 2]
```diff
@@ -78,20 +79,181 @@ static int gnix_verify_eq_attr(struct fi_eq_attr *attr)
...
+		ret = -FI_EINVAL;
+		goto cleanup_q_entry;
+	}
+
+	q_entry->len = len;
+	q_entry->buf = buf;
+	q_entry->type = event;
+	q_entry->flags = flags;
+
+	fastlock_acquire(&gnix_eq->lock);
```
Comment (bturrubiates) [inline prov/gni/src/gnix_eq.c:219]: I'm not sure how to handle the errrors for fastlock_init, fastlock_acquire, and fastlock_relese. They return regular errno codes and there isn't a conversion function for converting to fabric codes. Should I just log the output of strerror and return -FI_EBUSY? I checked around the project and the others that use fastlock\* don't check the return status. I'm not sure if it's important to be checking or not...

@hppritcha @sungeunchoi

**Comment by hppritcha** [comment]
I think this is ready to be merged.

**Comment by bturrubiates** [comment]
Did you review the new changes?

**Comment by hppritcha** [comment]
looks good,  merging.

**Pull request merged**